chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/ProductVersionController.java

```diff
@@ -154,4 +167,80 @@
+            throw new InvalidArgumentsException(msgUtil.getMessage("version.split.missingNewListing"));
+        }
+        if (splitRequest.getOldVersion() == null || splitRequest.getOldVersion().getVersionId() == null) {
+            throw new InvalidArgumentsException(msgUtil.getMessage("version.split.missingOldVersion"));
+        }
+        if (splitRequest.getOldListings() == null || splitRequest.getOldListings().size() == 0) {
+            throw new InvalidArgumentsException(msgUtil.getMessage("version.split.missingOldVersionListings"));
+        }
+        if (versionId.longValue() != splitRequest.getOldVersion().getVersionId().longValue()) {
+            throw new InvalidArgumentsException(msgUtil.getMessage("version.split.argumentMismatch",
+                    versionId, splitRequest.getOldVersion().getVersionId()));
+        }
+
+        HttpHeaders responseHeaders = new HttpHeaders();
+        ProductVersionDTO oldVersion = pvManager.getById(splitRequest.getOldVersion().getVersionId());
+        ProductVersionDTO newVersion = new ProductVersionDTO();
+        newVersion.setVersion(splitRequest.getNewVersionVersion());
+        newVersion.setProductId(oldVersion.getProductId());
+        newVersion.setDeveloperId(oldVersion.getDeveloperId());
+        List<Long> newVersionListingIds = new ArrayList<Long>();
+        for (CertifiedProduct requestNewVersionListing : splitRequest.getNewListings()) {
+            newVersionListingIds.add(requestNewVersionListing.getId());
+        }
+
+        ProductVersionDTO newVersionFromSplit = pvManager.split(oldVersion, newVersion, splitRequest.getNewVersionCode(),
+                newVersionListingIds);
+        responseHeaders.set("Cache-cleared", CacheNames.COLLECTIONS_LISTINGS);
+        SplitVersionResponse response = new SplitVersionResponse();
+        response.setNewVersion(new ProductVersion(newVersionFromSplit));
+        response.setOldVersion(new ProductVersion(oldVersion));
+
+        // find out which CHPL product numbers would have changed (only
+        // new-style ones)
+        // and add them to the response header
+        List<CertifiedProductDetailsDTO> possibleChangedChplIds = cpManager.getByVersion(newVersionFromSplit.getId());
+        if (possibleChangedChplIds != null && possibleChangedChplIds.size() > 0) {
+            StringBuffer buf = new StringBuffer();
+            for (CertifiedProductDetailsDTO possibleChanged : possibleChangedChplIds) {
+                if (!chplProductNumberUtil.isLegacy(possibleChanged.getChplProductNumber())) {
+                    if (buf.length() > 0) {
+                        buf.append(",");
+                    }
+                    buf.append(possibleChanged.getChplProductNumber());
+                }
+            }
+            responseHeaders.set("CHPL-Id-Changed", buf.toString());
+        }
+        return new ResponseEntity<SplitVersionResponse>(response, responseHeaders, HttpStatus.OK);
```

Comment:
Is it standard for us to return data via response headers?  If we need to this list back to the calling client, should it be part of the SplitVersionResponse object?

Reply:
There are a couple of other places where if something is edited the new-style CHPL ID could change and in all those places we send that information back in this CHPL-Id-Changed header. The UI looks for that and uses it to show one of those popups that tells you one or more CHPL IDs might have changed as a result of your action and changes could take a few minutes to show up. I just did this the same way... am open to discussion about a different/better way to do that, but I guess to answer your question it's "standard" in our CHPL code although I don't know if it's "standard" in a RESTful/best practices sense.